### PR TITLE
feat: join built-in function

### PR DESCRIPTION
**Pull request by zoldater**
## Description

The pull request introduces the `$join()` built-in function that returns a string value concatenating array elements with separator.

## Motivation

`sprintf` function replaces existing specifiers with the values from `arguments_list `. For this reason passing `$group()`, `$team()`, `$assignees()` or any array of arbitrary size as `arguments_list` could be correct.

Proposed change makes it possible to join arrays into string to replace only one format specifier.

## Example

- `$sprintf("These changes require approval by team 'a': %s", [$join($group("a"))])`

## Type of change

New feature (non-breaking change which adds functionality)

## How was this tested?

Unit tests.

## Checklist

- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works 
- [x] I have ran `task check -f` and have no issues

## Code review and merge strategy (ship/show/ask) 

 
 
- [x] Ask: this pull request requires a code review before merge

### CODE REVIEW

File: engine/transform_internal_test.go [hunk 1]
```diff
@@ -115,6 +115,46 @@ func TestTransformAladinoExpression(t *testing.T) {
...
+			wantVal: `$$join($team("a"), ", ")`,
+		},
+		"join assignees with empty separator": {
+			arg:     `$join($assignees())`,
+			wantVal: `$join($assignees(), " ")`,
+		},
+		"join assignees with non-empty separator": {
+			arg:     `$join($assignees(), ", ")`,
+			wantVal: `$join($assignees(), ", ")`,
+		},
```
Comment (marcelosousa) [inline engine/transform_internal_test.go:157]: The transform has a limitation that it doesn't work for nested $join statements. For example:

```
        "nested join with empty separators ": {
			arg:     `$join([$join(["a","b"]), "c"])`,
			wantVal: `$join([$join(["a","b"], " "), "c"], " ")`,
		},
```

**Review by marcelosousa — COMMENTED** [review]
Many thanks for this PR 🙌  I've left a comment related to the transform of the join built-in as it is not working for nested calls. I believe this is a broader issue of the way we are doing these transformations and should be addressed in a different issue. @ferreiratiago should I open an issue for that so that we can unblock this PR?

For the rest, it's looking good - thanks!

**Comment by marcelosousa** [comment]
Test coverage is good!

File: engine/transform_internal_test.go (hunk 1 above)
Comment (zoldater) [reply engine/transform_internal_test.go:157]: Thanks @marcelosousa !
I've added the test and fixed issue with recursion.

**Comment by zoldater** [comment]
Hi @ferreiratiago and @marcelosousa !
I fixed the issue with nested join calls and added proposed test to `engine/transform_internal_test.go`. So it should not be necessary to open new issue for that.

File: engine/transform.go [hunk 2]
```diff
@@ -73,6 +73,21 @@ func addDefaultHaveAllChecksRunCompleted(str string) string {
 	return r.ReplaceAllString(str, `$$haveAllChecksRunCompleted([], "")`)
 }
 
+func addDefaultJoinSeparator(str string) string {
+	r := regexp.MustCompile(`\$join\(((\[.*\])|(\$(team|group)\("[^"]*"\))|(\$assignees\(\)))(,\s*"([^"]*)")?\)`)
```
Comment (ferreiratiago) [inline engine/transform.go:77]: What's the reasoning behind `team`, `group`, and `assignees`? I may be missing something 😅

Isn't a simpler regex enough?
```suggestion
	r := regexp.MustCompile(`\$join\((\[.*\]|([^,]*))(,\s*"([^"]*)")?\)`)
```

If we have the literal values `team`, `group`, and `assignees`, we need to keep updating this transform every time there's a new built-in that returns a []string. For instance, [$filePaths()](https://github.com/reviewpad/reviewpad/issues/538).

**Review by ferreiratiago — CHANGES_REQUESTED** [review]

File: plugins/aladino/functions/join.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	if len(elementsArg.Vals) == 0 {
+		return aladino.BuildStringValue(""), nil
+	}
+	var clearVals []string
+	separatorArg := args[1].(*aladino.StringValue)
+	for _, val := range elementsArg.Vals {
+		switch v := val.(type) {
+		case *aladino.StringValue:
+			clearVals = append(clearVals, v.Val)
+		}
```
[the remark this answers is not in the document]
Comment (zoldater) [reply plugins/aladino/functions/join.go]: Done.

File: engine/transform.go (hunk 2 above)
Comment (zoldater) [reply engine/transform.go]: Thanks for pointing this out! Simplified the regex.

File: plugins/aladino/functions/join.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		return aladino.BuildStringValue(""), nil
+	}
+	var clearVals []string
+	separatorArg := args[1].(*aladino.StringValue)
+	for _, val := range elementsArg.Vals {
+		switch v := val.(type) {
+		case *aladino.StringValue:
+			clearVals = append(clearVals, v.Val)
+		default:
+			return nil, fmt.Errorf("join: invalid element of kind %v", v.Kind())
```
Comment (ferreiratiago) [inline plugins/aladino/functions/join.go:35]: We are very close. Just missing the unit test for this case.

**Review by ferreiratiago — CHANGES_REQUESTED** [review]
One unit test missing. Other than that it's perfect 🙌

File: plugins/aladino/functions/join.go (hunk 4 above)
Comment (zoldater) [reply plugins/aladino/functions/join.go:35]: Added `TestJoin_BadType`

**Review by ferreiratiago — APPROVED** [review]
Perfect. Thanks for your contribution @zoldater 🙏

**Pull request merged**